### PR TITLE
Add maxHivePartitions Strategy to avoid scanning excessive hive partitions on partitioned table

### PR DESCRIPTION
**Pull request by i7xh**
### _Why are the changes needed?_

The PR support this Strategy to avoid scan huge hive partitions  or unused necessary partition filter on partitioned table，It will abort the SQL query with thrown exception via scan hive table partitions exceed `spark.sql.watchdog.maxHivePartitions`, meanwhile show the related hive table partition struct to assist user to optimize sql according the given suggestion

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate
截屏2021-09-12 下午2 30 20

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @ulysses-you

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala [hunk 1]
```diff
@@ -18,8 +18,8 @@
 package org.apache.kyuubi.sql
 
 import org.apache.spark.sql.SparkSessionExtensions
-
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala:21]: unnecessary change

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.kyuubi.sql.KyuubiSQLExtensionException
+
+final class MaxHivePartitionExceedException(private val reason: String = "",
+                                            private val cause: Throwable = None.orNull)
+  extends KyuubiSQLExtensionException(reason, cause)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala:24]: it's readable to format code as:
```
final class MaxHivePartitionExceedException(
    private val reason: String = "",
    private val cause: Throwable = None.orNull)
  extends KyuubiSQLExtensionException(reason, cause)
```

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+              case Some(prunedPartitions) => if (prunedPartitions.size > maxHivePartition) {
+                throw new MaxHivePartitionExceedException(
+                  s"SQL job scan hive partition: ${prunedPartitions.size} " +
+                    s"exceed restrict of hive scan maxPartition $maxHivePartition \n" +
+                    s"You should optimize your SQL logical according partition structure" +
+                    s" or shorten query scope such as p_date, detail as below: \n" +
+                    s"Table: ${relation.tableMeta.qualifiedName} \n" +
+                    s"Owner: ${relation.tableMeta.owner} \n" +
+                    s"Partition Structure: " +
+                    s"${relation.partitionCols.map(_.name).mkString(" -> ")} \n")
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:52]: a little hard to read, can we use  ` """" xxx """" ` ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * 1 Check if scan exceed maxHivePartition
+ * 2 Check if Using partitionFilter on partitioned table
+ * This Strategy Add Planner Strategy after LogicalOptimizer
+ */
+case class MaxHivePartitionStrategy(session: SparkSession)
+  extends Strategy with SQLConfHelper {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_HIVEPARTITION) match {
+      case Some(maxHivePartition) => plan match {
+        case ScanOperation(_, _, relation: HiveTableRelation) =>
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:40]: nit: `case ScanOperation(_, _, relation: HiveTableRelation) if relation.isPartitioned =>`

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    s"You should optimize your SQL logical according partition structure" +
+                    s" or shorten query scope such as p_date, detail as below: \n" +
+                    s"Table: ${relation.tableMeta.qualifiedName} \n" +
+                    s"Owner: ${relation.tableMeta.owner} \n" +
+                    s"Partition Structure: " +
+                    s"${relation.partitionCols.map(_.name).mkString(" -> ")} \n")
+              } else {
+                Nil
+              }
+              case _ => throw new HivePartitionFilterUnusedException(
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:56]: shall we check if the hive table's total partition number is bigger or smaller than `maxHivePartition` ? I think we only need throw exception if its partition number is bigger than `maxHivePartition`.

**Review by ulysses-you — COMMENTED** [review]
thank you @i7xh , I think it is a useful config.  And not related this PR,  do we plan to support datasource table ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala (hunk 2 above)
Comment (i7xh) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala:25]: fixed

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala (hunk 5 above)
Comment (i7xh) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:56]: > thank you @i7xh , I think it is a useful config. And not related this PR, do we plan to support datasource table ?

In our scenario，we mainly facing huge hive table scan boring without partitioned key filter，support datasource table means other datasource such as ？

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:56]: Similar to the hive table`PruneHiveTablePartitions` and datasource table `PruneFileSourcePartitions`. The datasource table is also very common for user. We can improve this rule in a new PR that also support check datasource table if it's partitioned.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.kyuubi.sql.KyuubiSQLExtensionException
+
+final class MaxHivePartitionExceedException(
+  private val reason: String = "",
+  private val cause: Throwable = None.orNull)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala:24]: 4 indention for fields

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                    |or shorten query scope such as p_date, detail as below:
+                    |Table: ${relation.tableMeta.qualifiedName}
+                    |Owner: ${relation.tableMeta.owner}
+                    |Partition Structure: ${relation.partitionCols.map(_.name).mkString(" -> ")}
+                    |""".stripMargin)
+              } else {
+                Nil
+              }
+              case _ => val totalPartitions = session
+                .sessionState.catalog.externalCatalog.listPartitions(
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala:57]: `listPartitions` is a heavy option, I'd like to use `listPartitionNames` instead

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala (hunk 5 above)
Comment (i7xh) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxHivePartitionStrategy.scala]: > Similar to the hive table`PruneHiveTablePartitions` and datasource table `PruneFileSourcePartitions`. The datasource table is also very common for user. We can improve this rule in a new PR that also support check datasource table if it's partitioned.

Understood your mean, it's OK!

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala [hunk 8]
```diff
@@ -1584,4 +1584,40 @@ class KyuubiExtensionSuite extends QueryTest with SQLTestUtils with AdaptiveSpar
...
   }
+
+  test("test watchdog with scan maxHivePartitions") {
+    withTable("test", "temp") {
+      sql(
+        s"""
+           |CREATE TABLE test(i int)
+           |PARTITIONED BY (p int)
+           |STORED AS textfile""".stripMargin)
+      spark.range(0, 10000, 1).selectExpr("id as col")
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala:1595]: 10 rows is enough, otherwise we will produce 10000 * 10 rows data ..

**Review by ulysses-you — APPROVED** [review]

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala (hunk 8 above)
Comment (i7xh) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/KyuubiExtensionSuite.scala]: > 10 rows is enough, otherwise we will produce 10000 * 10 rows data ..

Excessive force～

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.kyuubi.sql.KyuubiSQLExtensionException
+
+final class MaxHivePartitionExceedException(
+    private val reason: String = "",
+    private val cause: Throwable = None.orNull)
+  extends KyuubiSQLExtensionException(reason, cause)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala:25]: style: new line

Comment (i7xh) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/KyuubiWatchDogException.scala]: Thanks!

**Comment by ulysses-you** [comment]
@i7xh  now, we upload the scalastyle output log https://github.com/apache/incubator-kyuubi/pull/1095, you can debug with that log by yourself

**Comment by ulysses-you** [comment]
thanks, merging to master

**Comment by ulysses-you** [comment]
@i7xh don't forget this https://github.com/apache/incubator-kyuubi/pull/1086#discussion_r707103617 : )

**Comment by ulysses-you** [comment]
hi @i7xh  seems your git user.email is not your github email, so your github account can not be seen as [contributor](https://github.com/apache/incubator-kyuubi/graphs/contributors). Can your check it ?

**Comment by i7xh** [comment]
> hi @i7xh seems your git user.email is not your github email, so your github account can not be seen as [contributor](https://github.com/apache/incubator-kyuubi/graphs/contributors). Can your check it ?

OK, I have already add my company email in my account
